Update the Kubeflow ROADMAP

jlewi opened the pull request:
* Provide info about the 0.6 and 0.7 releases
* Start to provide some clarity around 1.0
* The 1.0 items are largely just my oppinions and will likely go through
  several rounds of iterations as people weigh in and we start to build
  consensus.
* I think its useful to put a stake in the ground both with time and what
  we expect to deliver.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/kubeflow/3824)
<!-- Reviewable:end -->

jlewi commented:
/assign @abhi-g 
@ddutta @kkasravi @jbottum @vkoukis @theadactyl  @aronchick PTAL. 

In particular please take a look at the parts about 1.0. I'd like to get the ROADMAP updated reasonably soon so I suggest on this PR we focus on getting consensus on the high level bits. 

Specifically, do we all agree 

1. That we want to aim for 1.0 in Q4
1. That in order to achieve 1.0 in Q4 we will need to pair down 1.0 to its essentials and push important non-critical pieces to future releases?

kkasravi commented on ROADMAP.md:
```diff
@@ -13,79 +13,117 @@ for the issues tagged "area/X.Y.Z".
+   * v1beta1 for KfDef
+     * Create a more consistent and clean API for describing Kubeflow deployments
+     * Handle plugins and platforms consistently
+   * Clean up kfctl flags and semantics
```
@jlewi @abhi-g @ddutta 
A suggestion is we add 'accelerated training and inference support for deep learning models'.

We are prototyping this with an on-prem cluster so that a data scientist just runs a model 
and we have overlays in profile and tfoperator that injects device info that is optimal for that model.
We would also like to surface some of the device metrics up through the Application.

The value-add from matching a model (bert, MLP, transformer-lt) with hardware without requiring the data scientist know these details would be very compelling. This - for example - could be a profile overlay that could be prepackaged for common models or extended by vendors that are deploying kubeflow on known hardware and are focusing on specific models within their DS team.

jlewi replied:
Do we have specific deliverables for the 0.7; 1.0 time frame that we should list here?

kkasravi replied:
yes
- adding device overlays in manifests/profile that can be included in the namespace and job operators

jbottum commented:
We might want consider working backwards from what makes a component "1.0" i.e. what are the requirements for a "1.0 quality" doc, tutorial, install, upgrade, Unit Test, E2E test plan, Engineering Rules.    Can we put a process (especially automated processes) in place to deliver quality on the core components for the subsequent releases.

jlewi commented:
@jbottum I agree with all of that; what about that suggestion should we capture in the ROADMAP?

jlewi commented:
@jbottum and @kkasravi could you take a look at the questions I left you?

kkasravi commented:
@jlewi I updated above (https://github.com/kubeflow/kubeflow/pull/3824#discussion_r312646444)

jlewi commented:
@kkasravi I added a bullet in 0.7 about hardware acceleration.

kkasravi commented:
> @kkasravi I added a bullet in 0.7 about hardware acceleration.

@jlewi thank you!

jlewi commented:
@jbottum @abhi-g @kkasravi Could you folks LGTM this if this looks good to you.
We can continue to iterate and improve and follow on PRs but I think it would be good to get an updated ROADMAP published.

abhi-g commented:
thank you Jeremy for putting this together. 
/lgtm

jlewi commented:
@jbottum @kkasravi any final thoughts?

kkasravi commented:
We don't have much in the way of how an Application CR is used as the basic composition block inside KfDef. I think there are aspects around discoverability, interoperability, cleanup, health.

kkasravi commented:
Also do we want to say anything about system components and their ability to run within isolated namespaces, or anything about alignment with multi-tenant?

jbottum commented:
this is great work.    thanks for updating.   I dont want to slow up the process.   I would like to have a place holder for Persistent Volume Management via a Central Dashboard UI.    Note - Vangelis is on vacation and would like to get his input.

jbottum commented:
As for the definition of 1.0 quality definition, I saw the doc from Sarah Maddox and liked some pieces of that.  I was at a new user today, and their home-grown team / process is having Kubeflow stability issues (K8s is hard and this is even harder).   they also have security concerns.   Maybe we already have this listed but we might want "Best Practices for installation, operations and secure operations."

jbottum commented:
/lgtm

jlewi commented:
@jbottum I added a 0.7 section "Data management" and raised your point about persistent volume management via central dashboard UI

@kkasravi in 1.0 I have a bullet

> We will only support a single shared Kubeflow deployment per Kubernetes cluster

Does that cover multi-user support?

@kkasravi 

> We don't have much in the way of how an Application CR is used as the basic composition block inside KfDef. I think there are aspects around discoverability, interoperability, cleanup, health.

Added a bullet about monitoring in 0.7.

PTAL

jlewi commented:
/test all

jlewi commented:
Most recent test failure is

```
build github.com/kubeflow/kubeflow/bootstrap/v3/cmd/kfctlClient: cannot load github.com/kubeflow/kubeflow/bootstrap/cmd/bootstrap/app: cannot find module
providing package github.com/kubeflow/kubeflow/bootstrap/cmd/bootstrap/app
```

Looks like #3883

kkasravi commented:
/lgtm

ddutta commented:
/lgtm

On Tue, Aug 13, 2019 at 12:35 PM Kam Kasravi <notifications@github.com>
wrote:

> /lgtm
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubeflow/kubeflow/pull/3824?email_source=notifications&email_token=AAGWD6QPBLYQW77LWOKL2ILQEMEJRA5CNFSM4IJNJRLKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4GXU2A#issuecomment-520977000>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGWD6THIUCJW6WGW5QOCZTQEMEJRANCNFSM4IJNJRLA>
> .
>


-- 
-Debo~

jlewi commented:
@abhi-g @kkasravi @jbottum Could I get another LGTM please; I had to rebase to fix the tests.

kkasravi commented:
/lgtm

jlewi commented:
/approve

jbottum commented:
/lgtm
